src/once_cell.rs

```diff
@@ -101,3 +110,69 @@ impl<T> GILOnceCell<T> {
         Ok(())
     }
 }
+
+/// Interns `text` as a Python string and stores a reference to it in static storage.
+///
+/// A reference to the same Python string is returned on each invocation.
+///
+/// # Example: Using `intern!` to avoid needlessly recreating the same Python string
+///
+/// ```
+/// use pyo3::intern;
+/// # use pyo3::{pyfunction, types::PyDict, PyResult, Python};
+///
+/// #[pyfunction]
+/// fn create_dict(py: Python<'_>) -> PyResult<&PyDict> {
+///    let dict = PyDict::new(py);
+///    //             👇 A new `PyString` is created
+///    //                for every call of this function.
+///    dict.set_item("foo", 42)?;
+///    Ok(dict)
+/// }
+///
+/// #[pyfunction]
+/// fn create_dict_faster(py: Python<'_>) -> PyResult<&PyDict> {
+///    let dict = PyDict::new(py);
+///    //               👇 A `PyString` is created once and reused
+///    //                  for the lifetime of the program.
+///    dict.set_item(intern!(py, "foo"), 42)?;
+///    Ok(dict)
+/// }
+/// ```
+#[macro_export]
+macro_rules! intern {
+    ($py: expr, $text: expr) => {{
```

Comment:
Using `expr` means that the macro can capture values from its environment, which is an issue similar to the generics one:
```rust
pub fn foo<'a>(py: Python<'a>, thing: &'a str) -> &'a PyString {
    intern!(py, thing)
}

fn main() {
    Python::with_gil(|py| {
        let a = foo(py, "blah");
        let b = foo(py, "hmm");
        assert!(a.is(b));
    });
}
```

But we can fix that with a function, I think:
```rust
       // This function scope stops the macro from being able to capture variables.
        fn retrieve_static(py: $crate::Python<'_>) -> &$crate::types::PyString {
           INTERNED
            .get_or_init(py, || {
               $crate::conversion::IntoPy::into_py(
                   $crate::types::PyString::new(py, $text),
                    py,
                )
            })
            .as_ref(py)
        }
        retrieve_static($py)
```
which will prevent it from capturing its environment because functions cannot access the scope in which they're defined.

Reply:
Pushed.